### PR TITLE
Update ts configs, add diff-based linter.

### PR DESCRIPTION
**Pull request by goodov**
Sec-review to add `tsx` https://github.com/brave/reviews/issues/1575

Changes in this PR:

1. Add Prettier config for all files in the repository so it can be used natively via VSCode extension and as a part of `npm run lint`.
2. Update lint packages to most recent versions.
3. Add `lint-staged` integration to support diff linting with arbitrary executables. This will allow us to perform full study validation as a part of a single `npm run lint` call. `lint-staged` can also be integrated into `pre-commit` or `pre-push` git hooks, but after some experiments I decided to postpone this integration. 
4. Update `tsconfig` and `eslint` configs to run type checks on `src/`.

### CODE REVIEW

File: src/.eslintrc.js [hunk 1]
```diff
@@ -2,32 +2,28 @@
...
 // License, v. 2.0. If a copy of the MPL was not distributed with this file,
 // You can obtain one at https://mozilla.org/MPL/2.0/.
+
 module.exports = {
   settings: {
     react: {
       version: 'detect',
     },
   },
   extends: ['standard-with-typescript', 'prettier', 'plugin:react/recommended'],
```
Comment (goodov) [inline src/.eslintrc.js:12]: `prettier` stays here to disable `eslint` rules that `prettier` covers.

File: src/tsconfig.json [hunk 2]
```diff
@@ -7,13 +7,19 @@
     "allowJs": true,
     "moduleResolution": "node",
     "baseUrl": ".",
-    "typeRoots": [
-      "../node_modules/@types",
-    ],
+    "typeRoots": ["../node_modules/@types"],
+    "jsx": "react-jsx",
```
Comment (atuchin-m) [inline src/tsconfig.json:11]: It `src/web` specific thing, isn't it? We shouldn't use React outside web.

**Review by atuchin-m — DISMISSED** [review]

File: package.json [hunk 3]
```diff
@@ -3,47 +3,51 @@
...
     "build:tracker": "tsc -p src/finch_tracker/tsconfig.json",
     "build:web": "webpack -c src/web/webpack.config.js --mode=production",
-    "build": "npm run build:proto && npm run build:web && npm run build:tracker",
-    "serve:dev": "webpack serve -c src/web/webpack.config.js --hot --mode=development"
+    "clean": "tsx src/scripts/clean.ts",
+    "lint": "tsx src/scripts/lint.ts",
+    "serve:dev": "webpack serve -c src/web/webpack.config.js --hot --mode=development",
+    "test": "jest --config=src/jest.config.js",
+    "tracker": "node src/finch_tracker/build/finch_tracker/main.js",
+    "typecheck": "tsc --noEmit"
```
Comment (petemill) [inline package.json:15]: Will this be run as a blocker for merging? Otherwise there is no type-checking for the node code.

**Review by petemill — DISMISSED** [review]
I think this is a good solution, as long as `npm run typecheck` is used as a workflow to prevent merging code that fails to compile.

**Review by atuchin-m — APPROVED** [review]

**Pull request merged**